Flink: Create CatalogTestBase for migration to JUnit5

vinitpatni opened the pull request:
Base class creation of FlinkCatalogTestBase for the migration to JUnit5 in regards to https://github.com/apache/iceberg/issues/9079

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/FlinkCatalogTestBaseJU5.java:
```diff
@@ -0,0 +1,143 @@
+import org.junit.jupiter.api.io.TempDir;
+
+@ExtendWith(ParameterizedTestExtension.class)
+public abstract class FlinkCatalogTestBaseJU5 extends TestBase {
```
we can just call this `CatalogTestBase`

vinitpatni replied:
ack

nastra commented:
can you also please convert `TestMetadataTableReadableMetrics` to JUnit5 and use `CatalogTestBase` as its new base class?

vinitpatni commented:
I tested this implementation with couple of subclasses and it seems to be working fine. The difference that I observed here as compared to junit4 in terms of test execution. First test is triggered with all the supplied parameters in serial order then only it proceeds to next test in the sequence.

vinitpatni commented:
> can you also please convert `TestMetadataTableReadableMetrics` to JUnit5 and use `CatalogTestBase` as its new base class?

ack

nastra commented:
> I tested this implementation with couple of subclasses and it seems to be working fine. The difference that I observed here as compared to junit4 in terms of test execution. First test is triggered with all the supplied parameters in serial order then only it proceeds to next test in the sequence.

Can you include the subclasses you tested in this PR and convert them to JUnit5 in a separate commit?

vinitpatni commented:
Thanks for the review, @nastra   I have added following changes

- Rename FlinkCatalogTestBaseJU5 to CatalogTestBase
- Added Junit5 testcase and AssertJ style for TestFlinkCatalogDatabase

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,88 +29,93 @@
-        "Database should not already exist",
-        validationNamespaceCatalog.namespaceExists(icebergNamespace));
-
+    Assertions.assertThat(validationNamespaceCatalog.namespaceExists(icebergNamespace))
```
it's ok to have this statically imported

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,88 +29,93 @@
-    Assert.assertTrue(
-        "Database should still exist",
-        validationNamespaceCatalog.namespaceExists(icebergNamespace));
+
```
unnecessary newline here and in a bunch of places below

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,88 +29,93 @@
-    Assert.assertFalse(
-        "Namespace should not already exist",
-        validationNamespaceCatalog.namespaceExists(icebergNamespace));
+    Assumptions.assumeFalse(
```
please use AssertJ `Assumtions`: `assumeThat(...).as(...).isFalse()`

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -121,166 +126,194 @@ public void testDropNonEmptyNamespace() {
-        Assert.assertEquals("Should have 1 database", 1, databases.size());
-        Assert.assertEquals(
-            "Should have db and default database", "db", databases.get(0).getField(0));
+        Assertions.assertThat(databases.size()).as("Should have 1 database").isEqualTo(1);
```
```suggestion
        assertThat(databases).hasSize(1);`
```

nastra replied:
the advantage of this is that the full collection will be printed when the assertion ever fails. Please also update the other places to this style

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -121,166 +126,194 @@ public void testDropNonEmptyNamespace() {
-          databases.stream().anyMatch(d -> Objects.equals(d.getField(0), "db")));
+      Assertions.assertThat(databases.stream().anyMatch(d -> Objects.equals(d.getField(0), "db")))
+          .as("Should have db database")
+          .isTrue();
```
rather than using isTrue() / isFalse, we typically want to use a more fluent way of asserting this: `assertThat(databases).anyMatch(d -> Objects.equals(d.getField(0), "db"))`

vinitpatni replied:
ack

vinitpatni commented:
Addressed Following review comments

- Using static imports
- Removing redundant newline 
- Using Assumptions from Assertj library

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,227 @@
-
-    Assert.assertEquals(
-        "Namespace should have expected prop value", "value", nsMetadata.get("prop"));
+    assertThat(nsMetadata.get("prop"))
```
better to use `assertThat(nsMetadata).containsEntry("prop", "value")` as that will print the content of the map if the assertion ever fails

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,227 @@
-        "Namespace should have expected location",
-        "file:" + location.getPath(),
-        nsMetadata.get("location"));
+    assertThat(nsMetadata.get("location"))
```
same as above: better to use `.containsEntry(..)`

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,227 @@
-    Assert.assertFalse(
-        "Default metadata should not have custom property", defaultMetadata.containsKey("prop"));
-
+    assertThat(defaultMetadata.containsKey("prop"))
```
`assertThat(...).doesNotContainKey(..)`

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,227 @@
-
-    Assert.assertEquals(
-        "Namespace should have expected prop value", "value", nsMetadata.get("prop"));
+    assertThat(nsMetadata.get("prop"))
```
same as above

nastra reviewed (commented):
I've left a few minor comments, can you also please convert `TestMetadataTableReadableMetrics` and include it in this PR? In a follow-up PR we should then include all the other tests that currently extent `FlinkCatalogTestBase` and convert them to use `CatalogTestBase`.

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/CatalogTestBase.java:
```diff
@@ -0,0 +1,143 @@
+  protected boolean isHadoopCatalog;
+
+  @Parameters(name = "catalogName={0} baseNamespace={1}")
+  static List<Object[]> parametrs() {
```
```suggestion
  protected static List<Object[]> parameters() {
```

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/CatalogTestBase.java:
```diff
@@ -0,0 +1,143 @@
+  protected Namespace icebergNamespace;
+  protected boolean isHadoopCatalog;
+
+  @Parameters(name = "catalogName={0} baseNamespace={1}")
```
```suggestion
  @Parameters(name = "catalogName={0}, baseNamespace={1}")
```

vinitpatni replied:
ack

vinitpatni commented:
- Addressed Review Comments for CatalogTestBase and TestFlinkCatalogDatabase
- Moved parameters method to subclass of CatalogTestBase as there can be only single parameterProviders which will be defined by subclass extending  CatalogTestBase
- Adding Junit 5 conversion and AssertJ style for TestMetadataTableReadableMetrics and TestFlinkCatalogTablePartitions

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -18,270 +18,241 @@
-  public TestFlinkCatalogDatabase(String catalogName, Namespace baseNamespace) {
-    super(catalogName, baseNamespace);
+  @Parameters(name = "catalogName={0}, baseNamespace={1}")
+  static List<Object[]> parameters() {
```
I think it was correct the way it was before (by having it defined in `CatalogTestBase`). Subclasses that need different parameterization will just override this method. That's also how it is done in https://github.com/apache/iceberg/pull/9368

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:
```diff
@@ -40,28 +42,22 @@
-  @Parameterized.Parameters(name = "catalogName={0}, baseNamespace={1}")
-  public static Iterable<Object[]> parameters() {
+  @Parameters(name = "catalogName={0}, baseNamespace={1}")
+  static List<Object[]> parameters() {
```
no need to define this here, this should be at the super class

vinitpatni commented:
- Revert as per original behaviour and moved parameters to CatalogTestBase and overrided the parameters in subclass as per need

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,217 @@
+    assertThat(validationNamespaceCatalog.namespaceExists(icebergNamespace))
+        .as("Namespace should exist")
+        .isTrue();
+    assertThat(sql("SHOW TABLES").size()).as("Should not list any tables").isEqualTo(0);
```
```suggestion
    assertThat(sql("SHOW TABLES")).isEmpty();
```

vinitpatni replied:
ack

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/TestFlinkCatalogDatabase.java:
```diff
@@ -29,259 +32,217 @@
-          databases.stream().anyMatch(d -> Objects.equals(d.getField(0), "db")));
+      assertThat(databases)
+          .anyMatch(d -> Objects.equals(d.getField(0), "db"))
+          .as("Should have db database");
```
the `.as()` part needs to come before the `.anyMatch`

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:
```diff
@@ -76,7 +72,7 @@ protected TableEnvironment getTableEnv() {
   }
 
-  @Rule public TemporaryFolder temp = new TemporaryFolder();
+  @TempDir Path temp;
```
this can be private (unless there's a subclass that might use this)

nastra commented on flink/v1.18/flink/src/test/java/org/apache/iceberg/flink/source/TestMetadataTableReadableMetrics.java:
```diff
@@ -134,8 +130,7 @@ private Table createPrimitiveTable() throws IOException {
 
-    DataFile dataFile =
-        FileHelpers.writeDataFile(table, Files.localOutput(temp.newFile()), records);
+    DataFile dataFile = FileHelpers.writeDataFile(table, Files.localOutput(temp.toFile()), records);
```
this should be doing `File.createTempFile("junit", null, temp.toFile())` instead of `temp.toFile()`, similar to how it was done in #9341. The same needs to be done in L151

vinitpatni replied:
ack

nastra commented:
@vinitpatni just a few minor things, but should be ready to go once those are addressed

vinitpatni commented:
- Addressed Review comments

vinitpatni commented:
@nastra Thanks for merging my changes. I have also completed Junit 5 conversion and AssertJ style for TestFlinkCatalogTable and TestFlinkMetadataTable. Should I created Separate PR or follow up on this one ?

nastra commented:
Just create a new PR

vinitpatni commented:
@nastra Please find this new PR. https://github.com/apache/iceberg/pull/9381
